File: src/main/python/Advent2023/day_03.py
"""
https://adventofcode.com/2023/day/3
"""
import re
from collections import defaultdict

from src.main.python.utilities import LOGGER, timing


def adj_parts(a: int, b: int):
    """
    find all parts adjacent to a point (including the point)
    :param a:
    :param b:
    :return:
    """
    for a_ in a-1, a, a+1:
        for b_ in b-1, b, b+1:
            yield a_, b_


def parse(lines, gears=False):
    """
    parse the text, find all symbols (not including digits or .s)
    and yield either values adjacent to the symbol, or if the
    value is adjacent to a '*' and exactly one other is as well,
    yield those products.
    :param lines:
    :param gears:
    :return:
    """
    gear_dict = defaultdict(list)
    symbol_set = {
        (i, j)
        for j, line in enumerate(lines)
        for i, val in enumerate(line.replace(r'\n', ''))
        if not val.isdigit() and val != '.'
    }
    for j, line in enumerate(lines):
        for nums in re.finditer(r'\d+', line):
            for i in range(nums.start(), nums.end()):
                intersection = set(adj_parts(i, j)).intersection(symbol_set)
                if intersection:
                    n = int(nums.group(0))
                    for sym in intersection:
                        x, y = sym
                        if lines[y][x] == '*':
                            gear_dict[(x, y)].append(n)
                    if not gears:
                        yield n
                    break
    if gears:
        for v in gear_dict.values():
            if len(v) == 2:
                yield v[0] * v[1]


@timing
def part_one(lines) -> int:
    return sum(parse(lines))


@timing
def part_two(lines) -> int:
    return sum(parse(lines, gears=True))


if __name__ == '__main__':
    with open('../../resources/Advent2023/Day03.txt', 'r') as f:
        input_ = f.read().split('\n')
    LOGGER.info(part_one(input_))  # 528799   // 0.0093 sec
    LOGGER.info(part_two(input_))  # 84907174 // 0.0089 sec

File: src/main/scala/Advent2023/Day03.scala
/*
https://adventofcode.com/2023/day/3
*/

package Advent2023

import scala.util.matching.Regex
import scala.collection.mutable

import utilities.Timing

object Day03 {
  private def adjParts(a: Int, b: Int): Seq[(Int, Int)] = {
    for {
      a_ <- Seq(a - 1, a, a + 1)
      b_ <- Seq(b - 1, b, b + 1)
    } yield (a_, b_)
  }

  def parse(input: String, gears: Boolean = false): Int = {
    val numPattern: Regex = "\\d+".r
    val lines = input.split("\n")
    val symbolSet = lines.zipWithIndex.flatMap {
      case (line, j) => line.zipWithIndex.collect {
        case (char, i) if !char.isDigit && char != '.' => (i, j)
      }
    }.toSet
    val gearMap = mutable.Map[(Int, Int), mutable.ListBuffer[Int]]()
    val partNumSum = mutable.ListBuffer[Int]()

    lines.zipWithIndex.foreach { case (line, j) =>
      numPattern.findAllMatchIn(line).foreach { nums =>
        val n = nums.matched.toInt
        val intersectionFound = (nums.start until nums.end).exists { i =>
          val intersection = adjParts(i, j).toSet.intersect(symbolSet)
          intersection.nonEmpty && {
            intersection.foreach { case sym@(x, y) =>
              if (lines(y)(x) == '*') {
                if (gearMap.contains(sym)) gearMap(sym) += n else gearMap(sym) = mutable.ListBuffer(n)
              }
            }
            true
          }
        }
        if (intersectionFound && !gears) {
          partNumSum += n
        }
      }
    }
    if (gears) {
      gearMap.values.collect { case v if v.length == 2 => v.product }.sum
    }
    else partNumSum.sum
  }

  def part1(input: String): Int =
    parse(input)

  def part2(input: String): Int =
    parse(input, gears = true)

  def main(args: Array[String]): Unit =
    val data: String = io.Source.fromResource("Advent2023/Day03.txt").mkString
    part1(data)
    println(Timing.timing(part1(data)))  // 528799   // 0.024264736 s
    println(Timing.timing(part2(data)))  // 84907174 // 0.028974002 s
}
